Refactor code to support multiple virtio transports at runtime

slp opened the pull request:
Currently, virtio transports for each device are determined with
architecture dependent build time conditionals. This isn't the ideal
solution, as virtio transports aren't exactly tied to the host's
architecture.

For example, aarch64 VMs do support both PCI and MMIO devices, and
after the recent introduction of the microvm machine type, that's also
the case for x86_64.

This patch extends each device that supports multiple transports with
a VirtioTransport field, so users of the library can manually specify
a transport for each device. To avoid breaking the compatibility, if
VirtioTransport is empty a behavior equivalent to the legacy one is
achieved by checking runtime.GOARCH and Config.Machine.Type.

Keeping support for isVirtioPCI/isVirtioCCW in qmp.go is a bit
tricky. Eventually, the hot-plug API should be extended so callers
must manually specify the transport for the device.

Signed-off-by: Sergio Lopez <slp@redhat.com>

devimc commented on qemu/qmp.go:
```diff
@@ -1166,7 +1201,7 @@ func (q *QMP) ExecutePCIVFIODeviceAdd(ctx context.Context, devID, bdf, addr, bus
 	args := map[string]interface{}{
 		"id":       devID,
-		"driver":   Vfio,
+		"driver":   "vfio-pci",
```
changing constants with hard-coded strings, I'd prefer to keep constants

devimc commented on qemu/qemu.go:
```diff
@@ -87,32 +89,73 @@ const (
+	switch runtime.GOARCH {
+	case "amd64", "386":
+		switch config.Machine.Type {
+		case "microvm":
```
could you define a constant for this machine type?

devimc commented:
@slp your patch doesn't compile in kata

```
# github.com/kata-containers/runtime/vendor/github.com/intel/govmm/qemu
../vendor/github.com/intel/govmm/qemu/qemu_arch_base.go:23:2: Virtio9P redeclared in this block
	previous declaration at ../vendor/github.com/intel/govmm/qemu/qemu.go:77:26
../vendor/github.com/intel/govmm/qemu/qemu_arch_base.go:26:2: VirtioSerial redeclared in this block
	previous declaration at ../vendor/github.com/intel/govmm/qemu/qemu.go:80:30
../vendor/github.com/intel/govmm/qemu/qemu_arch_base.go:29:2: VirtioNet redeclared in this block
	previous declaration at ../vendor/github.com/intel/govmm/qemu/qemu.go:68:27
../vendor/github.com/intel/govmm/qemu/qemu_arch_base.go:29:27: undefined: VirtioNetPCI
../vendor/github.com/intel/govmm/qemu/qemu_arch_base.go:47:2: undefined: VirtioNetPCI
../vendor/github.com/intel/govmm/qemu/qemu_arch_base.go:50:2: undefined: VirtioBlockPCI
../vendor/github.com/intel/govmm/qemu/qemu_arch_base.go:68:6: method redeclared: NetDeviceType.QemuNetdevParam
	method(NetDeviceType) func(*NetDevice, *Config) string
	method(NetDeviceType) func() string
../vendor/github.com/intel/govmm/qemu/qemu_arch_base.go:68:24: NetDeviceType.QemuNetdevParam redeclared in this block
	previous declaration at ../vendor/github.com/intel/govmm/qemu/qemu.go:525:6
../vendor/github.com/intel/govmm/qemu/qemu_arch_base.go:89:6: method redeclared: NetDeviceType.QemuDeviceParam
	method(NetDeviceType) func(*NetDevice, *Config) DeviceDriver
	method(NetDeviceType) func() DeviceDriver
../vendor/github.com/intel/govmm/qemu/qemu_arch_base.go:89:24: NetDeviceType.QemuDeviceParam redeclared in this block
	previous declaration at ../vendor/github.com/intel/govmm/qemu/qemu.go:556:6
../vendor/github.com/intel/govmm/qemu/qemu_arch_base.go:50:2: too many errors
```

slp commented:
@devimc Ouch, I forgot to include the deletions in the commit. Also, I've just noticed I was working in the vendorized version of govmm, which still didn't have the tests in place.

Next week I'll fix the tests and address your suggestions about using constants.

slp commented:
@devimc I've fixed the tests and replaced the strings with constants. I've tested the library standalone and as part of Kata, but only in x86_64. I've been careful trying to avoid breaking s390x, but couldn't test it there.

Please let me know if you need any other changes.

alicefr commented:
@slp at first glance looks good to me. Thanks for the refactoring :) I like it

devimc commented:
@slp  some tests are failing

alicefr commented:
@slp might you create a non-merging PR in kata runtime with your changes? In this way, we can test them for the other architectures.

slp commented:
@devimc I fixed the lint errors, and added an arch-dependant package variable named "s390xTestOverride" to override runtime.GOARCH, so CI tests with s390x_test will still run on x86_64.

@alicefr Sure, done: https://github.com/kata-containers/runtime/pull/2359

markdryan commented:
I'm not that familiar with the qemu.go side of things, but I can't see any obvious problems.  If this commit doesn't break kata it's fine with me.

alicefr commented:
I think we could even get rid of the files `qemu/qemu_arch_base.go` and `qemu/qemus390x.go `. We have since a couple of month travis support for Z. So we could even let the test run on s390x and get rid of `s390x_test` tag. It's a little out of scope of this PR, but it will simplify it. Let me do a quicḱ PR and let's see.

markdryan commented:
@slp I'm afraid this now needs to be rebased as I've merged @alicefr travis patch.

slp commented:
Hi, so things on the kata-runtime front look fine (see https://github.com/kata-containers/runtime/pull/2359), and this PR is already rebased after the travis patch, so perhaps we can move it forward?

grahamwhaley commented:
@sboeuf @sameo - any chance you can take a look over this, given your experience I think in this area? thx!

devimc commented:
@slp looks like this needs a rebase :sweat_smile:

sameo commented on qemu/qemu.go:
```diff
@@ -87,32 +100,113 @@ const (
+	VfioPCI DeviceDriver = "vfio-pci"
+
+	// VfioCCW is the vfio driver with CCW transport.
+	VfioCCW DeviceDriver = "vfio-pci"
```
I think you want `vfio-ccw` here.

slp replied:
Ouch.

sameo commented on qemu/qemu.go:
```diff
@@ -299,6 +404,21 @@ func (fsdev FSDevice) QemuParams(config *Config) []string {
 	return qemuParams
 }
 
+// deviceName returns the QEMU device name for the current combination of
```
Nit: s/`QEMU device name`/`QEMU shared filesystem device name`/

slp replied:
Good catch!

sameo commented on qemu/qemu.go:
```diff
@@ -87,32 +100,113 @@ const (
+			switch config.Machine.Type {
+			case Microvm:
+				return false
+			}
```
I would separate the Microvm addition/support into a separate commit, making it clear that the first one is only about supporting multiple transports.

slp replied:
Yeah, makes sense.

slp commented:
@sameo I've applied your suggestions and rebased the patch. Please let me know if you find some other issue. Thanks!

slp commented:
@sameo ping? (sorry to bother you, but I'd prefer avoiding a new rebase, as its a large refactor and those are quite expensive, thanks!)

sameo commented:
@markdryan Are you okay with that PR?
